Hollow Knight: AP Website Tracker

mrkssr opened the pull request:
Please format your title with what portion of the project this pull request is
targeting and what it's changing.

ex. "MyGame4: implement new game" or "Docs: add new guide for customizing MyGame3"

## What is this fixing or adding?
A tracker for Hollow Knight on the AP website.

## How was this tested?
Local game hosting and creating different settings with text client item getting.

## If this makes graphical changes, please attach screenshots.
Multitracker overview over all available HK games.
In the charms section are rows just created for players who have either randomized charms or charm notches.
If a option is not randomized it will be shown with a dash on a grayish background. (there are some exceptions like eggs because they seem to be send to the AP tracker anyway - in that case that are just a few and not all found in the seed)
![HK_Multitracker1](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/8e39c8e7-46d0-4a4f-82f3-070987120985)

Handling of the skills
![HK_Multitracker3](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/2cd9944a-757f-4e03-9a04-44fb766a20e3)
![HK_Multitracker2](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/41ec6e87-a674-4b0d-8faa-e4d8ca1bf250)

Handling of the progressive charms.
![HK_Multitracker4](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/b9d59f20-5404-4937-bc47-3f4b7086f3e0)

View on a single player tracker.
![HK_Tracker1](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/608ded0e-40e3-466a-8def-2bc985ecc010)

The single player tracker just shows icons to options which were randomized. Not randomized stuff will be hidden / not generated for the page tracker.
![HK_Tracker2](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/5f0ed3d5-382d-4165-ac84-fb65a0c1adcd)

In the rare case someone chooses just options which are not handled by the tracker (i.e. Geo Rocks or Lore Tablets) there will just be shown a message. There seems not to be a way to avoid the tracker at all in this case therefore a short notification.
![HK_Tracker3](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/7f35acfb-e86c-48cd-b32c-044b264beb11)

BadMagic100 commented:
Going to leave this one for @ThePhar to review as I do not use the web tracker, don't really see a downside to have it but not going to rush to review it either and web is not my specialty. Looks neat though

ThePhar commented:
Assigned myself to review after work.

BadMagic100 commented:
@qwint I guess this is your problem now fyi

qwint commented on WebHostLib/static/styles/tracker__HK.css:
```diff
@@ -0,0 +1,260 @@
+    background-size: cover;
+    background-repeat: no-repeat;
+    background-position: center;
+} */
```
i assume this was commented while you were developing and can just be deleted now?

mrkssr replied:
Oh yes. I'll remove that.

qwint commented on WebHostLib/templates/multitracker__HK.html:
```diff
@@ -0,0 +1,478 @@
+    "Progressive Heart"             : ["Fragile_Heart", "Fragile_Heart", "Unbreakable_Heart"],
+    "Progressive Greed"             : ["Fragile_Greed", "Fragile_Greed", "Unbreakable_Greed"],
+    "Progressive Strength"          : ["Fragile_Strength", "Fragile_Strength", "Unbreakable_Strength"],
+    "Progressive Grimmchild"        : ["Grimmchild1", "Grimmchild1", "Grimmchild2", "Grimmchild3", "Grimmchild4"],
```
grimmchild doesn't work as a progressive item in AP
the difference between grimmchild1 and grimmchild2 is 2 tells the game that it should start upgraded so the player doesn't need the grimkin flames, where 1 tells the game to start in 1 form and force the player to upgrade it with the flames in the multiworld

we could still use the different icons for whichever is in the itempool / is sent, but they are not progressive

BadMagic100 replied:
I would keep a single icon, barring some plando nonsense there will only ever be 1 grimmchild in the seed

mrkssr replied:
Sometimes there is Grimmchild 1 and sometimes 2. I could rewrite it in code to just Grimmchild. It is guaranteed that there is just on Grimmchild at a time?

mrkssr replied:
The progressive is there more for the visual icon. So when Grimmchild1 is in pool then we get that icon, if it's two we get the other. Progression is here more because of using the existing stuff as workaround, independend of progressive behaviour or not in game.

Exempt-Medic replied:
Plando can cause both Grimmchilds to exist. But that can also cause, say, twenty Dream Nail items to exist or 5 version of a progressive item that only has two stages. I believe outside of plando there is only ever one Grimmchild

qwint replied:
https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/hk/__init__.py#L305
yes, either 1 or 2 will be added to the itempool, never both (without things like plando that we shouldn't need to cover for)
regardless it sounds like you understand they don't actually act like progression items, so whichever icons you chose to display is fine by me

qwint reviewed (changes requested):
I see the tooltip for Howling_Wraiths/Abyss_Shriek is "Progressive Shriek" but our item group for the two is "Scream"
I suggest following that same naming convention so we don't need to change it when we move to actual Progressive names

Elevator pass should be default on like focus/swim

Why have Progressive Dream Nail and Dreamgate? Dreamgate is the 2nd of the 3 Progressive Dream Nail stages

I'd want the horizontal (cloak/cdash) and vertical (claw/monarch wings) to be grouped together instead of staggered like they are currently

I'm not sure I like the progressive items showing as either the icon for the early form or a checkmark if fully gathered, but I'm not certain what I want it changed to,
My current ideas are: 
* always show the icon, for all stages
* show some fraction (i.e. 1/2 then 2/2 progressive fireballs)
* some partial checked icon for stage 1/2, but that would not work for 3 stage progressive items like dreamnail

is there any other webtrackers that have had to handle this? how do they display?

for the per player visual tracker, is there a way to make the background not pure black? It's hard to read 
in addition: some of the icons are so greyscale already that the difference between 'off' and 'on' is hard to see, I think a per-icon background (like a square/circle behind the icon) that gets shaded in or not as well would help but I'm open to other suggestions
in addition, the per player visual tracker also shows one icon for cloak when none are collected, even if split cloak is on, can you make the number of icons consistent? (either by always making two icons and have some way to denote full shade cloak, or by compressing the split cloak into one icon ex. mothwing cloak with an R denoting right cloak only, a mothwing cloak with an F denoting full mothwing, a shade cloak with an L denoting left shade cloak with no right dash) I'm open to other designs that keep the icon count the same too.

Can the dreamer icons just be their masks instead of the full body icon? I feel those are more recognizable

mrkssr commented:
> I see the tooltip for Howling_Wraiths/Abyss_Shriek is "Progressive Shriek" but our item group for the two is "Scream" I suggest following that same naming convention so we don't need to change it when we move to actual Progressive names
> 

I can change that. Are the other group names in the world code so I could take a look for the other groups as well?

> Elevator pass should be default on like focus/swim
> 

Focus/Swim are just checkmarked by default in the multi tracker. So I can do that for Elevator Pass there as well. In the single player tracker the other both are shown just when the option is randomized.

> Why have Progressive Dream Nail and Dreamgate? Dreamgate is the 2nd of the 3 Progressive Dream Nail stages
> 

Didn't know that all three are connected together. So I'll change it to Dreamgate as second between the other both.

> I'd want the horizontal (cloak/cdash) and vertical (claw/monarch wings) to be grouped together instead of staggered like they are currently
> 

I don't understand how that is meant / how that is visually expected.

> I'm not sure I like the progressive items showing as either the icon for the early form or a checkmark if fully gathered, but I'm not certain what I want it changed to, My current ideas are:
> 
> * always show the icon, for all stages
> * show some fraction (i.e. 1/2 then 2/2 progressive fireballs)
> * some partial checked icon for stage 1/2, but that would not work for 3 stage progressive items like dreamnail
> 
> is there any other webtrackers that have had to handle this? how do they display?
> 

Base is the ALttP tracker. There we have the mail (three stages) and sword (four stages). There is shown just the last/current one in the progression. That looks in my opinion also cleaner than i.e. a 1/2 and so on.

> for the per player visual tracker, is there a way to make the background not pure black? It's hard to read in addition: some of the icons are so greyscale already that the difference between 'off' and 'on' is hard to see, I think a per-icon background (like a square/circle behind the icon) that gets shaded in or not as well would help but I'm open to other suggestions in addition, the per player visual tracker also shows one icon for cloak when none are collected, even if split cloak is on, can you make the number of icons consistent? (either by always making two icons and have some way to denote full shade cloak, or by compressing the split cloak into one icon ex. mothwing cloak with an R denoting right cloak only, a mothwing cloak with an F denoting full mothwing, a shade cloak with an L denoting left shade cloak with no right dash) I'm open to other designs that keep the icon count the same too.
> 

- I could change the pure black background. Would that still be a problem when there is i.e. a colored background behind each icon so I have to change both or would the black background be fine when every icon has circle around?
- The idea with the left/right icons for cloak was to see that i.e. left is collected, right is missing. But maybe that is not necessary anymore if that is a single one, marked as collected and the L or R as note added.
- Not sure if we need an extra indicator for full cloak. If it is shown as collected and there is no letter then it's the full one.

> Can the dreamer icons just be their masks instead of the full body icon? I feel those are more recognizable

I have used the icons from https://hollowknight.wiki. There they were not other available. I would take a look for another source and change them.

mrkssr commented:
That is how full tracker looks like in ALttP. The bow, mail, shield, sword and gloves are progressive items. The number for the sword is part of the icon.
The bottle can be obtained multiple times. That is comparable with vessels, mask shards or Grubs in the HK tracker.

![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/9a4b4249-ddfb-46d5-a680-35fbfa062065)

Exempt-Medic commented:
For the Dreamers, you could use their [map pins](https://hollowknight.wiki/w/File:Map_Pin_Watcher.png)

mrkssr commented:
Haven't found them before but that are the one I wanted use anyway. I'll replace the current one then with them. :)

qwint commented:
item groups are defined here:
https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/hk/Items.py#L27

the horizontal/vertical comment was about grouping these columns
![image](https://github.com/ArchipelagoMW/Archipelago/assets/4809984/21423dcf-b977-47d9-9f1b-2a6fb9f1b47a)
in both places
![image](https://github.com/ArchipelagoMW/Archipelago/assets/4809984/48c18afb-3556-43fc-b33a-d5fbc87e67dc)

as both horizontal (cloak, cdash) and then both vertical (claw, wings)
the current order is more how one would naturally get them in vanilla, but that's not the most natural when dealing with rando

mrkssr commented:
Done changes:
- Changed the icons for the dreamers.
- Added a mapping for header texts for progressive items like fireball. Like that the background could stay the same like it was done in ALttP or OoT. (Now there is shown in header i.e.: Fireball, Mothwing/Shade Cloak, Scream)
- Dreamnail and Dreamgate is now merged and also set to progression by order (1. Dreamnail, 2. Dreamgate, 3. Awoken Dreamnail)
- Elevator Pass is in multi tracker marked as collected when it's not randomized in the pool.
- Grid layout changed. It's now everything below instead of the wide screen.
- In multi and single player tracker is the skill order now cloak, cdash, claw, wings.

Still to do:
- background
- behaviour for (not) splitted cloak

Not planned to changed at the moment: Grimmchild
- Progression for Grimmchild is used to show the sent one in tracker. In that way there is no extra logic needed to implement.
- Grimmchild 3 and 4 could be removed when they are not obtainable at all and it's not a thing for the future.

(hope, didn't forget anything)

mrkssr commented:
- Background color changed, icons have now also a circle around to support the visual if the item is collected or not.
- Splitted Mothwing/Shade Cloak is now also a single icon.

mrkssr commented:
Full view of collected icons. (all sections are now also below each other)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/5df32291-dcbb-4173-aef2-a14b1c0426e4)

Full view of not collected icons.
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/d1010e9c-85f6-4fee-b0c7-6352f2f79a26)

Multi tracker (Elevator Pass is now shown as equipped, when it's not in the pool).
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/99060944-54bc-421b-be3e-f21c145d22c0)

When Elevator is in pool and not collected the multi tracker shows it properly.
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/53e4f85d-49b2-4de1-aff5-fa1661326705)

Splitted Cloak: Double Left Cloak, no Right Cloak = Left Shade Cloak
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/789e0984-be20-447a-a298-5d7b99b4d013)

Splitted Cloak: Right Cloak, no Left Cloak = Right Mothwing Cloak
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/1b6dc9e2-7867-443f-b011-1bc2af912d25)

Dreamgate is part of the Dreamnail chain.
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/965826c0-4619-480c-8dcf-7dc85cccd214)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/3cf6e292-d0c6-4f3c-af63-ac1269d8230c)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/5a917b88-9452-4162-86f8-fedabbacdcd3)

Vertical and horizontal movement skills besides each other. (multi and player tracker)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/36f09c28-8706-4fc3-bdbf-5391989f2459)
![image](https://github.com/ArchipelagoMW/Archipelago/assets/68022469/e8c28970-552d-47c5-8af4-c64a562ec70c)

qwint commented:
why did Progressive Dive get switched to "Quake"? the item group is "Dive"
https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/hk/Items.py#L35

other than that this looks great, thanks

mrkssr commented:
As I searched for Scream I have read somewhere Quake for Dive. Probably here.
https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/hk/ExtractedData.py#L6

Overlooked then obviously that it is dive as you've sent the reference. Is now changed back to "Dive".

qwint reviewed (approved):
mostly reviewed content of the new webtrackers, but also skimmed through the code and everything looked ok. There are some grammar issues in the comments but I felt they read fine as to not worry.

Exempt-Medic commented on WebHostLib/tracker.py:
```diff
@@ -2396,3 +2396,308 @@ def render_Starcraft2_tracker(tracker_data: TrackerData, team: int, player: int)
+        )
+
+    _multiworld_trackers["Hollow Knight"] = render_HK_multiworld_tracker
+    _player_trackers["Hollow Knight"] = render_HK_tracker
```
EoF newline
```suggestion
    _player_trackers["Hollow Knight"] = render_HK_tracker

```

Exempt-Medic commented on WebHostLib/templates/multitracker__HK.html:
```diff
@@ -0,0 +1,493 @@
+        </table>
+    </div>
+{% endfor %}
+{% endblock %}
```
EoF newline
```suggestion
{% endblock %}

```

Exempt-Medic commented:
This seems to be missing the Godtuner

qwint commented:
> This seems to be missing the Godtuner

good point, Godtuner is now required progression for exactly the godhome flowerquest goal so it should be tracked properly

mrkssr commented:
I'm not at home this week. I'll add the stuff either in between if there is some time or next week.
What is Godtuner doing for the new goal that it is necessary? Isn't it just blinking when a boss was defeated.

qwint commented:
> I'm not at home this week. I'll add the stuff either in between if there is some time or next week. What is Godtuner doing for the new goal that it is necessary? Isn't it just blinking when a boss was defeated.

The godseeker flower quest dialogue that is required for the godhome flower ending won't trigger unless you have godtuner

mrkssr commented:
Godtuner added

Multitracker
![image](https://github.com/user-attachments/assets/ee2c698b-f0f2-47f8-89cb-d3706bae9958)

Single tracker
![image](https://github.com/user-attachments/assets/229e2e2e-3a93-48cd-a2fb-3d8286b5f20f)

mrkssr commented:
Is it fine so far or should be changed something else?

qwint commented on WebHostLib/tracker.py:
```diff
@@ -2471,3 +2471,308 @@ def render_Starcraft2_tracker(tracker_data: TrackerData, team: int, player: int)
+            "multitracker__HK.html",
+            enabled_trackers=enabled_trackers,
+            current_tracker="Hollow Knight",
+            room=tracker_data.room,
```
```suggestion
            room=tracker_data.room,
            get_slot_info=tracker_data.get_slot_info,
```
seems that the multitracker crashes without this

mrkssr replied:
Interesting. Hadn't that issue and the value is also not used in the html part for what that segment is. The slot data are just used directly in the same method.
But yeah, let's add it. Doesn't hurt anyway.

qwint commented:
I was also noticing that the non-multitracker version of the HK Tracker doesn't seem to be refreshing, not sure where that problem is because I wasn't seeing any error/warning on my webhost log

mrkssr commented:
Template for that was the ALttP tracker. There is just one element with the class .inventory-grid, here we have one per item type. So a change was missing to update all elements. Fixed.

Mathx2 commented:
> Splitted Cloak: Double Left Cloak, no Right Cloak = Left Shade Cloak ![image](https://private-user-images.githubusercontent.com/68022469/338752434-789e0984-be20-447a-a298-5d7b99b4d013.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY1NDA0MjMsIm5iZiI6MTczNjU0MDEyMywicGF0aCI6Ii82ODAyMjQ2OS8zMzg3NTI0MzQtNzg5ZTA5ODQtYmUyMC00NDdhLWEyOTgtNWQ3Yjk5YjRkMDEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDIwMTUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE1MzI0Nzc5Njc2M2MxMDAzMDk4ZWFkZjdjMDFmZmIwZjZmZGVmMDM0MDY1NzgwYjFkY2M1NTA4NGViNjYyMjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xIc6S34r_9E53ABG6PD3VgoQnJATNugS0GkqrzZA4jo)
> 
> Splitted Cloak: Right Cloak, no Left Cloak = Right Mothwing Cloak ![image](https://private-user-images.githubusercontent.com/68022469/338752651-1b6dc9e2-7867-443f-b011-1bc2af912d25.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY1NDA0MjMsIm5iZiI6MTczNjU0MDEyMywicGF0aCI6Ii82ODAyMjQ2OS8zMzg3NTI2NTEtMWI2ZGM5ZTItNzg2Ny00NDNmLWIwMTEtMWJjMmFmOTEyZDI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDIwMTUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmYjg1NDViNzhmNGNmMTZkYWRlNGI5MTc0NGYyOGE1ODAyODY2ODllYzY5ZTE5ZWYxMzZhOGE1N2FmMTU0NjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PVC9xxW8qhuyBrRbPCY6bZq1XO8jcfmGtnS9Q28f0xU)
> 

How would you handle Splitted Cloak: Double Left Cloak, Single Right Cloak?

qwint commented:
2 left 1 right or 1 left and 2 right are full shade cloak
i'm not sure what the worry is
